### DISCUSSION TITLE
chap 7 질문 정리

### DISCUSSION THREAD

**Discussion by sihyung92** [JS-OS-study/hyokyung-and-students#23] [General]
chap 7 - deadlock

**Comment by pullingoff** [comment]
# 느낀 점

- 데드락을 엄청 멋진 방법으로 방지하거나 처리할 줄 알았는데 그냥 무시한다니 살짝 허무할 뻔 했지만 허무하지 않았다!
- banker's algorithm이 실제로 어떻게 계산되는지 궁금하다. 

# 문제

<img width="535" alt="데드락 할당 그래프" src="https://user-images.githubusercontent.com/50111853/156174600-025e875d-e0c9-4f48-a953-ba9285fdf053.png">

그림 출처: https://www.geeksforgeeks.org/resource-allocation-graph-rag-in-operating-system/
1. 위 그림은 데드락인가? 답과 그 이유를 함께 서술하시오.
2. 'A 자원을 가지고 있어야 B 자원도 받을 수 있다'고 조건을 두는 것은 데드락의 4대 발생조건 중 무엇을 방지하기 위함인가?
3. 프로세스가 자원을 사용하는 절차(4단계)는?

**Reply by 1chz** [reply]
1. 데드락이다. 총 자원이 3개이며 P1, P2, P3이 각각 하나의 자원을 점유하고 있는 상태에서 사이클이 발생하고 있다. 즉, 데드락의 필요 조건 4가지 (상호배제, 비선점, 점유와 대기, 순환 대기) 를 모두 만족한다.
2. hold and wait
3. 요청 - 할당 - 사용 - 해제

**Reply by sihyung92** [reply]
1. 데드락이다. -> P1, P2, P3의 주세요 사이클이 돌고있어서 hold & waiting이 발생하는 것을 볼 수 있다.
2. 점유 및 대기
3. ~~요청, 대기, 할당, 해제 (감으로 찍었습니닷)~~ 대기가 아니라 사용이었군요!!

**Reply by inseo24** [reply]
1. 데드락인지 잘 모르겠음.(아닐 거 같음) 각 프로세스가 자원을 점유하고 있는 상태에서 사이클이 발생한 거 같은데 프로세스 3이 R2를 들고 있긴 한데 다른 자원을 기다리고 있는지?를 모르겠다. 

2. hold and wait?

3. request - allocate - use - release

**Comment by 1chz** [comment]
# 느낀점

---

- 데드락이 무섭다
- 윈도우에서 작업하다 프로세스에서 `응답 없음` 이 발생하면 별 생각없이 강제종료 했는데, 물론 원인이야 다양하겠지만 이게 데드락으로 인해 발생할수도 있다는 것 같다
- 데드락이 커널에서 발생하면 `블루 스크린`이 뜨는 경우도 있다고 한다
- 회사에서 사용하는 데이터베이스중에 MSSQL이 있는데, 얘는 가끔 로그에 자기가 직접 데드락을 해결했다는 로그가 찍히던데 `데드락 탐지 및 회복 기법`을 채택한건가...?

# 자바로 구현해 본 데드락

```java
public class Deadlock {
    private static final Logger logger = Logger.getLogger(Deadlock.class.getName());

    public static void main(String[] args) {
        Integer lock1 = 1;
        Integer lock2 = 2;

        Thread t1 = new Thread(() -> {
            synchronized (lock1) {
                logger.info("t1 got lock1. now going to try to get a lock2");
                synchronized (lock2) {
                    logger.info("t1 got lock2");
                }
                logger.info("If this message is printed, it means that t1 has acquired lock2");
            }
        });

        Thread t2 = new Thread(() -> {
            synchronized (lock2) {
                logger.info("t2 got lock2. now going to try to get a lock1");
                synchronized (lock1) {
                    logger.info("t2 got lock1");
                }
                logger.info("If this message is printed, it means that t2 has acquired lock1");
            }
        });

        t1.start();
        t2.start();
    }
}
```

```shell
INFO: t1 got lock1. now going to try to get a lock2
INFO: t2 got lock2. now going to try to get a lock1
```

위의 로그가 기록 된 이후 프로그램이 무한히 종료되지 않았다

# 문제

---

- 대부분의 현대 운영체제에서 선택한 데드락 해결 기법과, 그 이유를 간단하게 서술하시오.

**Reply by pullingoff** [reply]
데드락이 무섭다 2

**Reply by pullingoff** [reply]
대부분의 현대 운영체제는 데드락 무시 기법을 선택했다. 왜냐면 데드락은 드물게 발생하기 때문에 데드락을 탐지하고 회피하는 것보다 그냥 사용자가 알아서 프로세스를 죽이고 다시 시작하게 하는게 더 효율적이기 때문이다.

**Reply by sihyung92** [reply]
무시하기 -> 데드락을 탐지하거나, 회피하려고 감시할때나, 예방한다고 짠돌이처럼 계산하고 있는 오버헤드가 더 크기 때문에! 데드락이 자주 발생하지 않는 운영체제 입장에선 해당 비용이 더 크므로 무시한다.

**Reply by inseo24** [reply]
데드락을 무시함. 사용자가 직접 프로세스를 죽이는 방법을 선택함. 데드락을 방지하는 것의 오버헤드가 방지하지 않는 것보다 크다고 판단한 듯.

**Comment by sihyung92** [comment]
# 느낀 점
- 기계의 입장에서 데드락은 어찌보면 당연한 일이고(조건이 갖추어졌을 때) 사람입장에선 답답한 일이니.. 그를 해소하기 위한 다양한 알고리즘과 방법론이 정립된 것이 인상적이었다. 그러나 결국 OS에서 채택한 해결법은 냅두고 인간이 해결한다는 게 참 아이러니하다.
- mysql db 커넥션 풀에서 데드락이 발생한 경험이 있었는데, 그 때가 떠올랐다.

# 문제
- 데드락이 발생했을 때 처리하는 방법으로는 ____ , ____ , ___________ , ____ 가 있다.

**Reply by 1chz** [reply]
예방, 회피, 탐지 및 회복, 무시

**Reply by pullingoff** [reply]
방지(예방), 회피, 탐지 및 recover, 무시

**Reply by inseo24** [reply]
예방, 회피, 탐지&회복, 무시

**Comment by inseo24** [comment]
<h3>느낀 점 & 궁금한 점 </h3>

- 저번에 학습한 동기화와 연결된 문제라고 생각했다.
- 멀티 스레드 환경에서 데드락이 발생할 가능성이 있을 거 같은데 ... 병렬 프로그래밍을 공부해볼까 하고 생각했다.
- 아직 deadlock을 경험한 적이 없었던 거 같다. (있었더라도 몰랐을 듯)
- 데드락 발생 조건을 공부하고 해결 방법에 대해 배우는데 결론적으로 현대 OS에선 이를 방관하는 선택이 최선이라는게 인상적이었다.


<hr />
<h3>문제</h3>

Q. 데드락을 예방하는 방법 중 데드락 필요 조건 1개를 만족되지 않게 하는 방법이 있다. 여기서 반드시 성립되어야 하는 조건은 _____ 이다.(이 조건은 건들 수 없음)
A. ________(한글, 영문 상관없음)

**Reply by 1chz** [reply]
상호 배제

> 근데 현실적으로 대부분 보장해줘야 하는건 맞겠지만 반드시라고 볼 수는 없을 것 같아요! 100%랑 99%는 다르지 않을까 싶습니다. 프린터처럼 공유하면 안되는 자원도 있겠지만, 캠 같은건 공유해도 될 것 같기도 하네요.

**Reply by pullingoff** [reply]
데드락 방지 기법 중 공유하면 안되는 자원의 경우 상호 배제가 성립되어야함!

**Reply by sihyung92** [reply]
상호 배제
